device/src/state_machines/session.rs

```diff
@@ -0,0 +1,632 @@
+            rx_window: val.rx_window,
+        }
+    }
+}
+
+pub struct WaitingForRx<R>
+where
+    R: radio::PhyRxTx + Timings,
+{
+    shared: Shared<R>,
+    session: SessionData,
+    confirmed: bool,
+    rx_window: RxWindow,
+}
+
+impl<'a, R> WaitingForRx<R>
+where
+    R: radio::PhyRxTx + Timings,
+{
+    pub fn handle_event<C: CryptoFactory + Default>(
+        mut self,
+        event: Event<R>,
+    ) -> (Device<R, C>, Result<Response, super::super::Error<R>>) {
+        match event {
+            // we are waiting for the async tx to complete
+            Event::RadioEvent(radio_event) => {
+                // send the transmit request to the radio
+                match self.shared.radio.handle_event(radio_event) {
+                    Ok(response) => match response {
+                        radio::Response::RxDone(_quality) => {
+                            if let Ok(parsed_packet) =
+                                lorawan_parse(self.shared.radio.get_received_packet(), C::default())
+                            {
+                                if let PhyPayload::Data(data_frame) = parsed_packet {
+                                    if let DataPayload::Encrypted(encrypted_data) = data_frame {
+                                        let session = &mut self.session;
+                                        if session.devaddr() == &encrypted_data.fhdr().dev_addr() {
+                                            let fcnt = encrypted_data.fhdr().fcnt() as u32;
+                                            if encrypted_data.validate_mic(&session.newskey(), fcnt)
+                                                && (fcnt > session.fcnt_down || fcnt == 0)
+                                            {
+                                                session.fcnt_down = fcnt;
+
+                                                // this is a sane unwrap because we checked the MIC already
+                                                let decrypted = encrypted_data
+                                                    .decrypt(
+                                                        Some(&session.newskey()),
+                                                        Some(&session.appskey()),
+                                                        fcnt,
+                                                    )
+                                                    .unwrap();
+
+                                                let mut copy = Vec::new();
+                                                copy.extend(decrypted.as_bytes());
+
+                                                // this is a sane unwrap because we know the decrypted
+                                                // payload is a valid DecryptedDataPayload
+                                                self.shared.data_downlink = Some(
+                                                    DecryptedDataPayload::new_from_decrypted_payload(
```

Comment:
My understanding is that the copy is created here in order to be able to reuse the `data_frame` bytes for other purposes and in order to be able to keep the downlink in a field. If that is correct, I am wondering what is the advantage of this implementation over one where the copying is done earlier and the parsing logic has bytes that are owned right from the start. My reasoning is that I am not too fond of the idea of exposing an interface to create `DecryptedDataPayload` from raw bytes. I believe it could cause some confusion and lead to some errors if people are not careful when they provide encrypted data and when unencrypted one. That is to a great extend the reason to have two separate types - encrypted and unencrypted one. For that reason, I would rather not have such a method available, unless it is crucial for some use case. Other than those changes, I don't see any blocking points.

Reply:
I think that's a great idea. Thanks!